### PR TITLE
feat: recoverable processing + merging

### PR DESCRIPTION
**Pull request by andrzejnovak**
We discussed in the past how it's bad when a big job crashes and many hours and core hours can be lost. Here's a draft of how we could go about solving that (ignore some of the merge logic, I used the `FutureHolder` class from my job merging PR because this needs a bit finer control than the `_futures_handler` generator, but it's not necessarily connected)

Workflow can look like this. 

```
run_instance = processor.Runner(....
)

filemeta = run_instance(samples, "Events", processor_instance=MyZPeak(), prepro_only=True)` 
# save it in a pickle or sth for reference

# passing existing pre-made chunks to the Runner instance will skip the prepro step
output, processed, metrics = run_instance(filemeta , "Events", processor_instance=MyZPeak())` 
# WorkItem is attached to the accumulator like metrics are
# if exception is raised, report it, but return futures that already finished

missing = set(filemeta) - set(processed)
# resubmit
output, processed, metrics = run_instance(list(missing), "Events", processor_instance=MyZPeak())` 
```

@nsmith-

### CODE REVIEW

**Comment by lgray** [comment]
image

File: coffea/processor/executor.py [hunk 1]
```diff
@@ -216,6 +216,42 @@ def __call__(self, items):
         return accumulate(items)
 
 
+class FuturesHolder:
+    def __init__(self, futures, refresh=2):
+        self.futures = set(futures)
+        self.merges = set()
+        self.completed = set()
```
Comment (nsmith-) [inline coffea/processor/executor.py:246]: Won't this hold a reference to all task futures until the processing has finished? In which case all the intermediate output stays alive longer than it needs to. This would be really bad for local processing.

File: coffea/processor/executor.py [hunk 2]
```diff
@@ -553,23 +596,94 @@ def __call__(
...
-                        disable=not self.status,
-                        unit=self.unit,
-                        total=len(items),
-                        desc=self.desc,
-                    ),
-                    accumulator,
-                )
-            finally:
-                gen.close()
+                pbar = tqdm(disable=not self.status, unit=self.unit, total=len(items),
```
Comment (nsmith-) [inline coffea/processor/executor.py]: Add context managers for the progress bars (I should have done that before)

File: coffea/processor/executor.py [hunk 3]
```diff
@@ -216,6 +216,42 @@ def __call__(self, items):
...
+                self.merges,
+                timeout=refresh,
+                return_when=concurrent.futures.FIRST_COMPLETED,
+            )
+            self.completed.update(completed)
+            self.done["merges"] += len(completed)
+
+    def fetch(self, N):
+        return [
+            self.completed.pop().result() for _ in range(N) if len(self.completed) > 0
```
Comment (nsmith-) [inline coffea/processor/executor.py]: `range(max(N, len(self.completed))`

**Review by nsmith- — COMMENTED** [review]
Some early comments, mostly I'm concerned about object lifetime. One of the first problems we had was holding onto futures longer than necessary and blowing up our memory, as discussed in #97

File: coffea/processor/executor.py (hunk 2 above)
Comment (andrzejnovak) [reply coffea/processor/executor.py]: that's there just to make the merging optional, nesting with with the conditional looks weird

File: coffea/processor/executor.py (hunk 1 above)
Comment (andrzejnovak) [reply coffea/processor/executor.py:246]: Well technically with merging the intermediate/finished futures should get cleaned up. OTOH the merge futures get queue only after the standard jobs, not sure if there's a way around that. 

The future itself should add negligible overhead, not sure how bad we are doing with multiple hists vs. merged hist.

Comment (andrzejnovak) [reply coffea/processor/executor.py:246]: Ok, added intermediate merging on the submit node when merge jobs are turned off. Should be functionally equivalent to current implementation

Comment (nsmith-) [reply coffea/processor/executor.py:246]: Futures hold internally a ref to the result, which is large (histograms)
But if the goal here is to only hold `completed` that are pending a merge job and `merges` only holds finished merges pending a final merge, then I think that's ok. We might be in danger of having too deep a queue here though, but maybe lets wait and see if it becomes a problem

**Comment by andrzejnovak** [comment]
The magic sauce seems to be a secondary executor (either spawned automatically or passed in) to run the merge jobs in parallel. 

I am not sure where we landed on having rich as a dependency, but rich progress bars make the code with multiple bars a bit neater https://github.com/andrzejnovak/coffea/pull/1/files

**Comment by NJManganelli** [comment]
pip made rich a vendored dependency for 22.0, so was it transitively via that?

**Comment by andrzejnovak** [comment]
Actually seems quite a bit faster this time around, two runs each, merge pool with 2 workers, which can probably run anywhere.
image
image

image
image

**Comment by lgray** [comment]
Can you replicate the work to be done a little bit and see how the memory scales?

**Comment by andrzejnovak** [comment]
image
 Same files though, so I am guessing the files are cached somehow so the time doesn't scale x10. Looking at `htop` while that happens, the mem consumption looks consistent throughout. I am running a slightly modded MyZPeak example with 4x60bins axes, so should be not entirely negligible.

**Comment by lgray** [comment]
Be careful with tests in windows - forking and multiprocessing is strange there.

**Comment by andrzejnovak** [comment]
Alright, I guess that makes it reviewable now. Specifically, I'd like to hear your thoughts on the added API and particularly for how the error state/processed items are returned back up by the executor(s). This is currently passed as a tuple, but it's not very elegant. I assume the reason the returns are formed like eg. `(accumulator, metrics)` as opposed to just returning the `wrapped_out` dict, is that somewhere it has to be immutable?

**Comment by lgray** [comment]
We should test this on a many-core (64+) machine and check the scaling in extremum as well.

**Comment by lgray** [comment]
We'll still run into some nasty issues with big histograms... hmm.

**Comment by andrzejnovak** [comment]
> We should test this on a many-core machine and check the scaling in extremum as well.

I can test this on a node up to 80 cores, but there's a lot of RAM available, so it's not necessarily stressing the code.

The next scale up is obv the parsl executor (either with Futures or another parsl dfk doing the merging), which should maybe be a separate PR.

**Comment by lgray** [comment]
Sure - the point is to observe trends rather than really stress anything.

**Comment by andrzejnovak** [comment]
Alright, this crept up in scope a bit. 
- New features (for futures/parsl executors)
  - `recoverable` - Optionally return currently processed chunks along with the error in case of failure, instead of raising directly
  - `merging` - Implement merging in batches (either in main process/in main executor/in separate executor)
    - On small test jobs at least as fast as master or faster 
    - Will particularly improve when many small chunks are being processed or when many workers are available
  - rich progress bars
- API changes - breaking
  - convert sets to list in metrics (it's annoying they are not be default json dumpable)
  - Executors return `(out, 0)` or `(out, error)` if `recoverable` is implemented
- API changes
  - Calling Runner type standardized - instead of a variable-length tuple, return dict
    - isolated into Runner().run(...)
    - current `Runner()()` remain unchanged - under the hood calls `run()` 
    - deprecate and switch in the future?

Ready to review. We can expand the merge logic to dask and overall use of rich progress bars in future PRs.

File: coffea/processor/executor.py [hunk 4]
```diff
@@ -1296,26 +1612,64 @@ def __call__(
...
+        ----------
+            fileset : dict
+                A dictionary ``{dataset: [file, file], }``
+                Optionally, if some files' tree name differ, the dictionary can be specified:
+                ``{dataset: {'treename': 'name', 'files': [file, file]}, }``
+            treename : str
+                name of tree inside each root file, can be ``None``;
+                treename can also be defined in fileset, which will override the passed treename
+            processor_instance : ProcessorABC
+                An instance of a class deriving from ProcessorABC
```
Comment (andrzejnovak) [inline coffea/processor/executor.py:1653]: fix docstring

File: coffea/processor/executor.py [hunk 5]
```diff
@@ -208,7 +215,7 @@ def __str__(self):
         return "reduce"
 
     def __call__(self, items):
-        items = list(items)
+        items = list(it for it in items if it is not None)
```
Comment (nsmith-) [inline coffea/processor/executor.py:232]: Now that this list is filtered, is there a case where `len(items)==0`? Since it would trigger the exception immediately following

File: coffea/processor/executor.py [hunk 6]
```diff
@@ -555,30 +623,152 @@ def __call__(
...
+        def processwith(pool, mergepool):
+            reducer = _reduce(self.compression)
+            FH = FuturesHolder(
+                set(pool.submit(function, item) for item in items), refresh=2
             )
+
+            nparts, minred, maxred = 5, 4, 100
+            if isinstance(self.merging, tuple):
+                if len(self.merging) == 3:
+                    nparts, minred, maxred = self.merging
```
Comment (nsmith-) [inline coffea/processor/executor.py]: Move above `pool.submit` and return an error if the format isn't valid

Comment (nsmith-) [reply coffea/processor/executor.py]: Best would be a dataclass `__post_init__` hook https://docs.python.org/3/library/dataclasses.html#post-init-processing

File: coffea/processor/executor.py [hunk 7]
```diff
@@ -555,30 +623,152 @@ def __call__(
...
+                    merged = None
+                    while len(FH.futures) + len(FH.merges) > 0:
+                        FH.update()
+                        merge_size = min(
+                            maxred, max(len(FH.completed) // nparts + 1, minred)
+                        )
+                        progress.update(
+                            p_id,
+                            advance=FH.done["futures"]
+                            - progress._tasks[p_id].completed,
```
Comment (nsmith-) [inline coffea/processor/executor.py]: Is there a public method on the rich progress bar?

File: coffea/processor/executor.py [hunk 8]
```diff
@@ -218,6 +225,41 @@ def __call__(self, items):
...
+                self.merges,
+                timeout=refresh,
+                return_when=concurrent.futures.FIRST_COMPLETED,
+            )
+            self.completed.update(completed)
+            self.done["merges"] += len(completed)
+
+    def fetch(self, N):
+        return [
+            self.completed.pop().result() for _ in range(min(N, len(self.completed)))
```
Comment (nsmith-) [inline coffea/processor/executor.py]: Since a future's exception would be raised from here (`f.result()`) then any other completed futures that were popped before getting to the broken one would get dropped from the recovery process, right? Can we let `fetch` return the future and later access its result, perhaps after an explicit check if it has an `f.exception()`?

File: coffea/processor/executor.py [hunk 9]
```diff
@@ -817,28 +1029,143 @@ def __call__(
...
-                    disable=not self.status,
-                    unit=self.unit,
-                    total=len(items),
-                    desc=self.desc,
-                ),
-                accumulator,
-            )
-        finally:
-            gen.close()
+            with rich_bar() as progress:
```
Comment (nsmith-) [inline coffea/processor/executor.py]: How much of this section is different from the `FuturesExecutor` impl? It looks quite similar, can we deduplicate? (maybe by putting the future update loop into FuturesHolder and passing it a callback to submit the processing and reducing task?)

File: coffea/processor/executor.py [hunk 10]
```diff
@@ -817,28 +1029,143 @@ def __call__(
...
+                                total=len(recovered),
+                            )
+                        ),
+                        self.compression,
+                    )
 
-        return accumulator
+                    return accumulate([_decompress(merged), accumulator]), e
+            else:
+                raise type(e)(str(e)).with_traceback(sys.exc_info()[2]) from None
```
Comment (nsmith-) [inline coffea/processor/executor.py]: Based on #650 this should just be `raise Exception(...) from e`

File: coffea/processor/executor.py [hunk 11]
```diff
@@ -901,20 +1228,14 @@ class Runner:
...
+    cachestrategy: Optional[
+        Union[Literal["dask-worker"], Callable[..., MutableMapping]]
+    ] = None  # fmt: skip
     processor_compression: int = 1
     use_skyhook: Optional[bool] = False
+    skyhook_options: Optional[Dict] = field(default_factory=dict)
     format: str = "root"
 
-    @staticmethod
-    def read_coffea_config():
```
Comment (nsmith-) [inline coffea/processor/executor.py]: Why is this deleted?

File: tests/test_processor.py [hunk 12]
```diff
@@ -117,7 +117,7 @@ def postprocess(self, accumulator):
         proc.process,
         None,
     )
-    assert out == {"itemsum": 45}
+    assert out == ({"itemsum": 45}, 0)
```
Comment (nsmith-) [inline tests/test_processor.py:120]: Based on the fact that `__call__` preserves the old behavior, why did this end up changing?

**Review by nsmith- — CHANGES_REQUESTED** [review]
As an overall comment, if you can sprinkle some typing in some of the new function signatures that would be helpful.

File: tests/test_processor.py (hunk 12 above)
Comment (andrzejnovak) [reply tests/test_processor.py:120]: Runner.__call__ preservers old behaviour, this is invoking the exec directly

File: coffea/processor/executor.py (hunk 11 above)
Comment (andrzejnovak) [reply coffea/processor/executor.py]: Hmm, something weird happened when I updated the branch

File: coffea/processor/executor.py (hunk 10 above)
Comment (andrzejnovak) [reply coffea/processor/executor.py]: 👍

File: coffea/processor/executor.py (hunk 9 above)
Comment (andrzejnovak) [reply coffea/processor/executor.py]: refactored

File: coffea/processor/executor.py (hunk 7 above)
Comment (andrzejnovak) [reply coffea/processor/executor.py]: there isn't but there's a different way to advance the progress

File: coffea/processor/executor.py (hunk 5 above)
Comment (andrzejnovak) [reply coffea/processor/executor.py:232]: I suppose not, but if `len(items)==0` then the original would throw just below anyway no? This is in place if a default new accumulator is being added to the merge, but can just be initiated as None

**Comment by andrzejnovak** [comment]
Instead of passing "prepro" to run, I tried to factor it out into a separate fcn, which makes more sense, but having to comply with the dynamic chunking there makes it a bit awkward (relies on both filemeta and chunk generator, so `run` needs to be able to take `fileset` as input.

**Comment by andrzejnovak** [comment]
@nsmith- Cleaned up the commits and also added rich.progress for Iterative. I'd say this is ready

**Comment by lgray** [comment]
Since we can't capture it very well in CI have we tried this PR at scale with all the executors?

**Comment by andrzejnovak** [comment]
> Since we can't capture it very well in CI have we tried this PR at scale with all the executors?

Reasonable scale for futures, full 2016 production of my analysis with parsl about 3 times. The speed there is dependent on how busy the disk, but in each it was faster or as fast as master

File: coffea/processor/executor.py [hunk 13]
```diff
@@ -1315,7 +1622,47 @@ def __call__(
             # v0.7.4. This fixes tests using maxchunks.
             fileset.reverse()
 
-        chunks = self._chunk_generator(fileset, treename)
+        return self._chunk_generator(fileset, treename)
+
+    def run(
+        self,
+        fileset: Union[Dict, List, str],
```
Comment (nsmith-) [inline coffea/processor/executor.py]: Generator now too?

File: coffea/processor/executor.py [hunk 14]
```diff
@@ -525,25 +695,65 @@ class FuturesExecutor(ExecutorBase):
...
+            Callable[..., concurrent.futures.Executor],
+            concurrent.futures.Executor,
+            bool,
+        ]
+    ] = None
+    recoverable: bool = False
+    merging: Union[bool, Tuple[int, int, int]] = False
+    nparts: int = 5
+    minred: int = 4
+    maxred: int = 100
```
Comment (nsmith-) [inline coffea/processor/executor.py]: I guess I was thinking to still have the merging as one arg. This is a bit of duplication of parameters but probably ok since its validated after construction

**Comment by lgray** [comment]
@nsmith-  you happy here?

**Review by nsmith- — APPROVED** [review]
Just a small type thing, not making it required but would be nice.

File: coffea/processor/executor.py (hunk 14 above)
Comment (andrzejnovak) [reply coffea/processor/executor.py]: seems requisite to making the parsing in the _post_init

**Comment by andrzejnovak** [comment]
@nsmith- seems we're good to go
